README.md
## feathersjs-serverless

[![npm version](https://badge.fury.io/js/feathersjs-serverless.svg)](https://badge.fury.io/js/feathersjs-serverless)
[![Build Status](https://travis-ci.org/adrianovalente/feathersjs-serverless.svg?branch=master)](https://travis-ci.org/adrianovalente/feathersjs-serverless)

> feathersjs connector to the Serverless framework ⚡️

Have you ever imagined building an application with [feathersjs](https://feathersjs.com/) and deploying it using the [Serverless Framework](https://serverless.com/)?

``` javascript
const feathers = require('@feathersjs/feathers')
const serverless = require('feathersjs-serverless')

const Todos = {
  find: async () => ([{
    description: 'Build a nice application'
  }, {
    description: 'Deploy it using Serverless'
  }])
}

const app = serverless(feathers())
  .use('todos', Todos)

module.exports.handler = app.handler()
```


CHANGELOG.md
## CHANGELOG
### 0.3.1
- Bump dependencies

### 0.3.0
- Bump dependencies

### 0.2.0
- Use `resource` rather than `path` to find service

### 0.1.0
- Bump dependencies

### 0.0.3
- Add default empty object for query

### 0.0.2
- Initial Release


test/handler.test.js
const feathers = require('@feathersjs/feathers')
const serverlessApp = require('../src')

test('set/get api express-like', () =>
  expect(serverlessApp()
    .set('foo', 'bar')
    .get('foo'))
    .toEqual('bar')
)

test('integration test: happy path', async () => {
  const cb = jest.fn()
  const fooBar = [{ foo: 'bar' }]
  const app = serverlessApp(feathers())
    .use('hello', {
      find: async () => fooBar
    })
    .handler()

  await app({ resource: 'hello', httpMethod: 'GET' }, null, cb)

  expect(cb).toHaveBeenCalledWith(null, { body: JSON.stringify({ data: fooBar }) })
})


test/get-service.test.js
const feathers = require('@feathersjs/feathers')
const getService = require('../src/get-service')

const dummyService = { find: () => [] }
const feathersApp = feathers()
  .use('people', dummyService)
  .use('people/shirts', dummyService)

test('unknown service', () => {
  const { feathersId, service } = getService(feathersApp, '/foo')
  expect(feathersId).toEqual(null)
  expect(service).toBe(null)
})

test('people service', () => {
  const { feathersId, service } = getService(feathersApp, '/people')
  expect(feathersId).toEqual(null)
  expect(service).toBe('people')
})

test('people service with id', () => {
  const { feathersId, service } = getService(feathersApp, '/people/1234')
  expect(feathersId).toEqual('1234')
  expect(service).toBe('people')
})

test('people/shirts service', () => {
  const { feathersId, service } = getService(feathersApp, '/people/shirts')
  expect(feathersId).toEqual(null)
  expect(service).toBe('people/shirts')
})

test('people/shirts service with id', () => {
  const { feathersId, service } = getService(feathersApp, '/people/shirts/1234')
  expect(feathersId).toEqual('1234')
  expect(service).toBe('people/shirts')
})


src/index.js
const Proto = require('uberproto')

const getService = require('./get-service')
const getArgs = require('./get-args')
const { getFeathersMethod } = require('./methods')

module.exports = feathersApp => {
  const mixin = {
    set (key, value) {
      if (!this.variables) {
        this.variables = {}
      }

      Object.assign(this.variables, {
        [key]: value
      })

      return this
    },

    get (key) {
      return this.variables[key]
    },

    setup (func) {
      this.setupFunc = func
      return this
    },

    handler () {
      const self = this
      self.emit('handlerstarted')
      return async (event, context, cb) => {
        if (!self.setupPromise) {
          self.setupPromise = (typeof self.setupFunc === 'function')
            ? self.setupFunc(self)
            : Promise.resolve()
        }

        await self.setupPromise

        const {
          resource,
          httpMethod: method,
          body: bodyAsString
        } = event

        const query = event.queryStringParameters || {}

        const body = bodyAsString
          ? JSON.parse(bodyAsString)
          : {}

        const { service: serviceName, feathersId } = getService(self, resource)

        if (!serviceName || !self.service(serviceName)) {
          return cb(null, {
            statusCode: 404,
            body: JSON.stringify({ error: `Service not found: ${resource}` })
          })
        }

        const service = self.service(serviceName)
        const feathersMethod = getFeathersMethod(method, feathersId)

        if (!feathersMethod || !service[feathersMethod]) {
          return cb(null, {
            statusCode: 404,
            body: JSON.stringify({ error: `Method not allowed: ${method}` })
          })
        }

        const args = getArgs(feathersMethod, { query, feathersId, body })
        return service[feathersMethod](...args)
          .then(data => {
            return cb(null, {
              body: JSON.stringify({ data })
            })
          })
          .catch(e => {
            return cb(null, {
              statusCode: e.code || 500,
              body: JSON.stringify({ error: e.message })
            })
          })
      }
    }
  }

  return Proto.mixin(mixin, feathersApp)
}


src/handler.js
const Proto = require('uberproto')

const getService = require('./get-service')
const getArgs = require('./get-args')
const { getFeathersMethod } = require('./methods')

const serverlessApp = feathersApp => {
  const mixin = {
    set (key, value) {
      if (!this.variables) {
        this.variables = {}
      }

      Object.assign(this.variables, {
        [key]: value
      })

      return this
    },

    get (key) {
      return this.variables[key]
    },

    setup (func) {
      this.setupFunc = func
      return this
    },

    handler () {
      const self = this
      self.emit('handlerstarted')
      return async (event, context, cb) => {
        if (!self.setupPromise) {
          self.setupPromise = (typeof self.setupFunc === 'function')
            ? self.setupFunc(self)
            : Promise.resolve()
        }

        await self.setupPromise

        const {
          path,
          httpMethod: method,
          queryStringParameters: query,
          body: bodyAsString
        } = event

        const body = bodyAsString
          ? JSON.parse(bodyAsString)
          : {}

        const { service: serviceName, feathersId } = getService(self, path)

        if (!serviceName || !self.service(serviceName)) {
          return cb(null, {
            statusCode: 404,
            body: JSON.stringify({ error: `Service not found: ${path}` })
          })
        }

        const service = self.service(serviceName)
        const feathersMethod = getFeathersMethod(method, feathersId)

        if (!feathersMethod || !service[feathersMethod]) {
          return cb(null, {
            statusCode: 404,
            body: JSON.stringify({ error: `Method not allowed: ${method}` })
          })
        }

        const args = getArgs(feathersMethod, { query, feathersId, body })
        return service[feathersMethod](...args)
          .then(data => {
            return cb(null, {
              body: JSON.stringify({ data })
            })
          })
          .catch(e => {
            return cb(null, {
              statusCode: e.code || 500,
              body: JSON.stringify({ error: e.message })
            })
          })
      }
    }
  }

  return Proto.mixin(mixin, feathersApp)
}

module.exports.serverlessApp = serverlessApp
